### DISCUSSION TITLE
hw_02的幾個的思路(想自己想的不要點進來，雖然我也不算給答案

### DISCUSSION THREAD

**Discussion by ghost** [1101-GS4719-C/Python-Community-GS4719-C#129] [Ideas]
1. 在開始寫之前，我習慣先拉架構出來，最複雜的一定是畫*字號那部分，所以那部分先跳過，先把其他功能寫好。這次作業應該蠻明顯的整體架構就是會用到if做輸入值的判斷，所以可以先把這部分寫一寫。如果直接寫不太會，我覺得可以先把要做幾種事情列出來，再去想要怎麼分配到if/elif/else裡面。

2. 如果對 **def** 不太熟可以自己上網搜尋一下，印象中上課好像沒有特別提到(呃也有可能是我在恍神XD) 這裡[提供一篇](https://medium.com/ccclub/ccclub-python-for-beginners-tutorial-244862d98c18)，看不懂就去找其他篇
3. 再來是*字號的部分，如果毫無頭緒的話，一樣先用想的把要怎麼做的才能畫出那樣的圖?(先不用管語法，先想你需要甚麼功能就好，不管PYTHON有沒有這功能)。列完功能後，完整的確認一次用這些功能理論上可以做出作業，那就可以開始一步步去找/試/查python有哪些語法可以用，正式開始寫這段程式。我自己是習慣不跟前面拉的主架構一起寫，可以另外拉出一個區域寫def這一段，這樣就先不用管主架構的縮排那些，寫完跑沒問題再放回去就好。
我是習慣寫到哪查到哪，後面還沒用到的功能就等之後要寫了再去查可以用什麼語法，在一開始就一次查完全部我覺得可能會有查到的功能不適用在你目前寫的程式上的問題

需要的功能/語法EX.
- [ ] *
- [ ] " "
- [ ] 顯示在螢幕上
- [ ] 讓使用者輸入
- [ ] 重複:要顯示出很多層，每層之間好像有點規律，**有點重複感**，寫程式時不知道要幾層所以要能把使用者入的值放進去
- [ ] 「自己看敘述想」*字號從作業說明看是print字串出來的樣子，因為不可能事先把所有row數的字串打好print出來，又有重複性，所以可以推斷出每row字串的來源應該是對前or後row的字串做出改變，這樣需要用什麼功能呢?   提示:這個功能可能會遇到有些資料型態不能做到你想做的事，那就上網查換個資料型態看看吧!
...
4. 要找出怎麼寫row*字號位置的程式蠻難的，這部分我也建議沒想法的話先不要碰程式，先寫在紙上，觀察一下米字號出現的位置，還有他的位置與row的關係，一步步去想要怎麼達成，可以先用3~5row畫出來去想，數量小比較容易想出來
提示:有時候不一定要一次畫好，可以用多畫幾次的方式來試看看
5. 想不到完整的寫法沒關係，想過之後有一點想法了就開始試，有時候數學算式太抽象太麻煩，不夠直觀很難想，直接用程式試跑出結果直接看我覺得更有幫助
試跑的小技巧:為了讓自己知道程式跑到哪裡做了什麼，可以加一些額外的print，像是如果有迴圈跑到第幾圈，某個變數的值現在是多少，現在數學公式算出來的值是多少....等等，最後做完記得註解掉或刪掉
6.我不知道我的寫法是不是最簡潔的，所以不一定會用到我上面有打的功能喔...

字有點多，感謝看到這裡的人XD，我很菜請大家多多指教
不知道這樣會不會洩漏太多耶...如果不能po這種文可以跟我講一下，我再刪掉

**Comment by chia-kai-chang** [comment]
太感動了，看到同學這麼認真討論，真的很開心 (大叔笑)

**Reply by ghost** [reply]
平常都忘記要發文，趁有作業多少加一點發文次數XD

**Comment by vvvioletta** [comment]
同學你好，我想請教一下~如果要讓他每次都print出Enter the number of rows是不是應該要多放while loop直到遇到-1 break(但不確定這樣寫對不對)

**Reply by ghost** [reply]
摁我其實沒有用while(所以難怪寫起來感覺有點麻煩嗎XD)，你的意思是說想列出現在是在畫第幾行嗎?我是在for loop內寫一個COUNT變數初始值=0，然後寫count = count +1，再print(count)
顯示出來大概長這樣(第二行的start program是之前在寫整體架構時測試用的，忘記刪掉了)
![image](https://user-images.githubusercontent.com/90567166/138654968-90f68dfc-fee4-4ecf-8222-01593e6264bd.png)
不知道有沒有回答道你的問題XD，用while應該也可以
```
i = 0
while i < row:
    i += 1
    print(i)
    ...後面寫你print出一行*的程式碼
```

break印象中是要跳出迴圈在用的，好像不用用到也可以...?其實我跟while不太熟==有沒有其他大神幫忙解答一下?

**Reply by vvvioletta** [reply]
我是想說while loop裡面包for loop，如果遇到-1直接break，但如果遇到其他數字就進去for loop

**Reply by jerry850313** [reply]
那你的while判斷條件為何

**Reply by vvvioletta** [reply]
我用無限迴圈耶 因為感覺老師是希望他一直跑 跑到遇到-1為止

**Reply by jerry850313** [reply]
但是有-1的break條件
我覺得改成"while n!=-1"比較安全
再來就是迴圈數量太多
執行速度會降低

**Reply by jerry850313** [reply]
如果是要判斷 n=-1就跳出的話
用 if 比較適合

**Reply by ghost** [reply]
> 我是想說while loop裡面包for loop，如果遇到-1直接break，但如果遇到其他數字就進去for loop


我覺得用無限迴圈可以喔，因為他每次都要等使用者輸入後才會往下跑，所以應該不會有太多圈的問題(欸就程式的邏輯理論上是這樣，但實際上電腦是怎麼跑的會不會吃很多資源我就不太清楚了==.)

可以像類似下面這樣測試看看
```
i = 0
while 1:
    i += 1
    a = input("輸入")
    if a == "-1":
        break
    print("row")  #我用這個代表寫 * 的for迴圈
    print(i)  #檢查用
```

**Reply by jerry850313** [reply]
答案是會。
越多的判斷式意味著越多的邏輯閘，每一個邏輯閘都代表著時間成本。
這次作業其實是感覺不太出來的，但如果是好幾個project的程式碼串接，就會明顯感受到差異。
再來說點題外話，晶圓在設計的時候會考量幾個問題 :
> 1.面積越小越好
> 2.驅動電流越小越好
以上這兩點都跟邏輯閘數量息息相關，所以原則上邏輯閘數量必須要控制。
這也表示電腦跟人最大的差異點 : 我們都希望程式碼越少行越好，而電腦則是邏輯閘越少越好。

**Reply by ghost** [reply]
> 答案是會。 越多的判斷式意味著越多的邏輯閘，每一個邏輯閘都代表著時間成本。 這次作業其實是感覺不太出來的，但如果是好幾個project的程式碼串接，就會明顯感受到差異。 再來說點題外話，晶圓在設計的時候會考量幾個問題 :
> 
> > 1.面積越小越好
> > 2.驅動電流越小越好
> > 以上這兩點都跟邏輯閘數量息息相關，所以原則上邏輯閘數量必須要控制。
> > 這也表示電腦跟人最大的差異點 : 我們都希望程式碼越少行越好，而電腦則是邏輯閘越少越好。

同學你好，我想請問如果我寫的無限迴圈每圈都需要等待手動輸入才會繼續往下跑，這樣會有邏輯閘太多的問題嗎?(程式編寫的順序上我覺得應該不會，還是說電腦會因為我用無限迴圈就預先分配較多資源給迴圈?
謝謝!

**Reply by jerry850313** [reply]
會
因為它會回到邏輯閘之前，等你輸入

**Reply by ghost** [reply]
了解了，謝謝!

**Reply by jerry850313** [reply]
它比較像是由邏輯閘組成模組，再由模組執行你的程式的概念。每個迴圈都可以當作一個模組，巢狀迴圈則是模組包著模組的概念。所以需要輸入的迴圈就相當於這個模組執行之前，必須給他東西才會跑。

**Comment by jerry850313** [comment]
可以用while判斷式來判斷
我是直接用n<=0當條件
因為題目是只需要上三角形


n=1
while n>0:
    print("Enter the number of rows : ",end="")
    n=int(input())                        //這裡會把輸入的str轉成int
    printstar(n)                           //我把主要的部份放在function (printstar)裡面
print("End Program")               //當n不符合while條件的時候就會執行這行

會另外寫一個function是因為這樣的code比較好維護，把code分成一段一段比較容易知道bug在哪
